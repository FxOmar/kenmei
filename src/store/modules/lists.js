import { Message } from 'element-ui';
import * as userTags from '@/services/endpoints/UserTags';
import * as mangaEntries from '@/services/endpoints/v2/manga_entries';

// Can't access getter inside mutations, hence this has to be a plain function
export const getEntryIndex = (state, id) => state.entries.findIndex(
  (e) => e.id === id,
);

const state = {
  tags: [],
  entries: [],
  selectedEntries: [],
  entriesPagy: {},
  statuses: [
    { enum: 1, name: 'Reading' },
    { enum: 2, name: 'On Hold' },
    { enum: 3, name: 'Plan to Read' },
    { enum: 4, name: 'Completed' },
    { enum: 5, name: 'Dropped' },
  ],
  tagsLoading: false,
  entriesLoading: true,
  entriesUpdating: false,
  checkboxToggled: false,
};

const getters = {
  findEntryFromIDs: (state) => (ids) => state.entries.find(
    (entry) => ids.includes(entry.id),
  ),
  selectedEntriesIDs: (state) => state.selectedEntries.map((entry) => entry.id),
  selectionModeActive: (state) => state.selectedEntries.length > 0 && state.checkboxToggled,
};

const actions = {
  async getTags({ commit }) {
    const response = await userTags.index();
    const { status, data } = response;

    return status === 200 ? commit('setTags', data) : Message.error(data.error);
  },
  async getEntries({ commit }, { page, status, tagIDs, searchTerm, sort }) {
    commit('setEntriesLoading', true);

    const response = await mangaEntries.index(
      page, status, tagIDs, searchTerm, sort,
    );

    if (response.status === 200) {
      commit('setEntries', response.data.entries);
      commit('setEntriesPagy', response.data.pagy);
    } else {
      Message.error(response.data.error);
    }

    commit('setEntriesLoading', false);
  },
  async getEntriesPagy({ commit }, { page, status, tagIDs, searchTerm, sort }) {
    const response = await mangaEntries.pagyInfo(
      page, status, tagIDs, searchTerm, sort,
    );

    if (response.status === 200) {
      commit('setEntriesPagy', response.data);
    } else {
      Message.error(response.data.error);
    }
  },
};

const mutations = {
  setTags(state, data) {
    state.tags = data;
  },
  setEntries(state, data) {
    state.entries = data;
    state.selectedEntries = [];
  },
  setSelectedEntries(state, { entries, isCheckbox }) {
    state.checkboxToggled = isCheckbox;
    state.selectedEntries = entries;
  },
  setEntriesPagy(state, data) {
    state.entriesPagy = data;
  },
  addEntry(state, data) {
    state.entries.unshift(data);
  },
  addSelectedEntry(state, { entry, isCheckbox }) {
    state.checkboxToggled = isCheckbox;
    state.selectedEntries.push(entry);
  },
  updateEntry(state, data) {
    state.entries.splice(getEntryIndex(state, data.id), 1, data);
  },
  replaceEntry(state, { currentEntry, newEntry }) {
    state.entries.splice(getEntryIndex(state, currentEntry.id), 1, newEntry);
  },
  setTagsLoading(state, data) {
    state.tagsLoading = data;
  },
  setEntriesLoading(state, data) {
    state.entriesLoading = data;
  },
  setEntriesUpdating(state, data) {
    state.entriesUpdating = data;
  },
};

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
};
