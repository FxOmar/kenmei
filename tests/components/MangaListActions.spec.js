import Vuex from 'vuex';

import MangaListActions from '@/components/MangaListActions.vue';

import lists from '@/store/modules/lists';

const localVue = createLocalVue();

localVue.use(Vuex);

// TODO: Properly test this component
describe('MangaListActions.vue', () => {
  let store;
  let dashboardActions;

  beforeEach(() => {
    store = new Vuex.Store({
      modules: { lists: { namespaced: true, state: lists.state } },
    });

    dashboardActions = shallowMount(MangaListActions, {
      store,
      localVue,
      propsData: {
        selectedStatus: 1,
        selectedTagIDs: [],
        debouncedSearchTerm: '',
      },
      stubs: ['router-link'],
    });
  });

  it('renders component', async () => {
    expect(dashboardActions.html()).toMatchSnapshot();
  });
});
