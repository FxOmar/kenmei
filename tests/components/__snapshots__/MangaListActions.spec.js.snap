// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangaListActions.vue renders component 1`] = `
<div class="mx-2 mb-5 sm_mx-5 flex justify-between flex-col sm_flex-row">
  <div class="flex flex-col w-full sm_flex-row">
    <el-select-stub value="1" autocomplete="off" multiplelimit="0" placeholder="Filter by status" valuekey="value" popperappendtobody="true" class="w-full sm_w-40">
      <el-option-stub value="-1" label="All"></el-option-stub>
      <el-option-stub value="1" label="Reading"></el-option-stub>
      <el-option-stub value="2" label="On Hold"></el-option-stub>
      <el-option-stub value="3" label="Plan to Read"></el-option-stub>
      <el-option-stub value="4" label="Completed"></el-option-stub>
      <el-option-stub value="5" label="Dropped"></el-option-stub>
    </el-select-stub>
    <el-select-stub value="" autocomplete="off" filterable="true" multiple="true" multiplelimit="0" placeholder="Filter by tags" valuekey="value" collapsetags="true" popperappendtobody="true" class="w-full mt-3 sm_mt-0 sm_ml-3 sm_w-48"><template></template></el-select-stub>
    <sort-dropdown-stub class="w-full mt-3 sm_mt-0 sm_ml-3 sm_w-40"></sort-dropdown-stub>
  </div>
  <div class="mt-3 text-center w-full float-right sm_mt-0 sm_text-left sm_w-64">
    <base-form-input-stub value="" type="text" placeholder="Input manga title"><template></template>
      <!---->
    </base-form-input-stub>
  </div>
</div>
`;
